File: backend/app/routes/resume.py
from fastapi import (
    APIRouter,
    UploadFile,
    File,
    HTTPException,
    Depends,
)

from sqlalchemy.orm import Session

import shutil
import os
import uuid


from app.services.resume_parser import extract_resume_text
from app.services.resume_analyzer import analyze_resume

from app.database import SessionLocal

from app.models.resume import ResumeResult

from app.auth import get_current_user





router = APIRouter()





UPLOAD_FOLDER = "uploads"


os.makedirs(

    UPLOAD_FOLDER,

    exist_ok=True

)






# =========================
# Database Dependency
# =========================

def get_db():


    db = SessionLocal()


    try:

        yield db


    finally:

        db.close()







# =========================
# Analyze Resume
# =========================


@router.post("/analyze")
async def analyze_resume_file(


    file: UploadFile = File(...),


    db: Session = Depends(get_db),


    current_user = Depends(get_current_user)


):


    extension = (

        file.filename

        .split(".")

        [-1]

        .lower()

    )




    if extension not in [

        "pdf",

        "docx"

    ]:


        raise HTTPException(

            status_code=400,

            detail="Only PDF and DOCX files are supported"

        )







    file_id = str(uuid.uuid4())



    file_path = (

        f"{UPLOAD_FOLDER}/"

        f"{file_id}_{file.filename}"

    )





    try:



        with open(

            file_path,

            "wb"

        ) as buffer:


            shutil.copyfileobj(

                file.file,

                buffer

            )







        text = extract_resume_text(

            file_path

        )






        if not text.strip():


            raise HTTPException(

                status_code=400,

                detail="Unable to extract resume text"

            )







        result = analyze_resume(

            text

        )








        # Save Resume With User ID

        resume_result = ResumeResult(


            user_id=current_user["id"],


            filename=file.filename,


            score=result["score"],


            skills=", ".join(

                result["skills"]

            ),


            missing_keywords=", ".join(

                result["missingKeywords"]

            ),


            strengths=", ".join(

                result["strengths"]

            ),


            suggestions=", ".join(

                result["suggestions"]

            )


        )






        db.add(resume_result)


        db.commit()


        db.refresh(resume_result)







        return {


            "id":

                resume_result.id,


            "filename":

                resume_result.filename,


            "score":

                resume_result.score,


            "skills":

                result["skills"],


            "missingKeywords":

                result["missingKeywords"],


            "strengths":

                result["strengths"],


            "suggestions":

                result["suggestions"],


            "created_at":

                resume_result.created_at


        }







    except HTTPException:

        raise



    except Exception as error:


        db.rollback()


        raise HTTPException(

            status_code=500,

            detail=str(error)

        )







    finally:


        if os.path.exists(file_path):

            os.remove(file_path)









# =========================
# Resume History
# =========================


@router.get("/history")
def resume_history(


    db: Session = Depends(get_db),


    current_user = Depends(get_current_user)


):


    results = (


        db.query(ResumeResult)


        .filter(

            ResumeResult.user_id == current_user["id"]

        )


        .order_by(

            ResumeResult.created_at.desc()

        )


        .all()


    )







    return [



        {


            "id":

                item.id,



            "filename":

                item.filename,



            "score":

                item.score,



            "skills":

                item.skills.split(", ")

                if item.skills

                else [],



            "missingKeywords":

                item.missing_keywords.split(", ")

                if item.missing_keywords

                else [],



            "strengths":

                item.strengths.split(", ")

                if item.strengths

                else [],



            "suggestions":

                item.suggestions.split(", ")

                if item.suggestions

                else [],



            "created_at":

                item.created_at



        }



        for item in results



    ]

File: frontend/src/services/resume.ts
import { apiFetch } from "./api";


// =========================
// Resume Analysis Interface
// =========================

export interface ResumeAnalysis {

  id?: number;

  filename: string;

  score: number;

  skills: string[];

  missingKeywords: string[];

  strengths: string[];

  suggestions: string[];

  created_at?: string;

}





// =========================
// Resume History Interface
// =========================

export interface ResumeHistoryItem {

  id: number;

  filename: string;

  score: number;

  skills: string[];

  missingKeywords: string[];

  strengths: string[];

  suggestions: string[];

  created_at: string;

}







// =========================
// Analyze Resume
// =========================

export async function analyzeResume(

  file: File

): Promise<ResumeAnalysis> {



  const formData = new FormData();


  formData.append(

    "file",

    file

  );






  const response = await apiFetch(

    "/api/resume/analyze",

    {

      method: "POST",

      body: formData,

    }

  );







  if(!response.ok){


    const error =

      await response.text();



    console.error(

      "Resume analysis error:",

      error

    );



    throw new Error(

      error ||

      "Resume analysis failed"

    );


  }








  const result =

    await response.json();







  return {


    id:

      result.id,



    filename:

      result.filename || "",



    score:

      Number(result.score) || 0,



    skills:


      Array.isArray(result.skills)

      ? result.skills

      : [],



    missingKeywords:


      Array.isArray(result.missingKeywords)

      ? result.missingKeywords

      : Array.isArray(result.missing_keywords)

      ? result.missing_keywords

      : [],



    strengths:


      Array.isArray(result.strengths)

      ? result.strengths

      : [],



    suggestions:


      Array.isArray(result.suggestions)

      ? result.suggestions

      : [],



    created_at:

      result.created_at || "",


  };


}









// =========================
// Resume History
// =========================

export async function getResumeHistory()

: Promise<ResumeHistoryItem[]> {



  const response = await apiFetch(

    "/api/resume/history",

    {

      method:"GET",

    }

  );







  if(!response.ok){



    const error =

      await response.text();



    console.error(

      "Resume history error:",

      error

    );



    throw new Error(

      error ||

      "Failed to load resume history"

    );


  }







  const result =

    await response.json();







  if(!Array.isArray(result)){


    return [];


  }








  return result.map(



    (item:any)=>({



      id:

        item.id,





      filename:

        item.filename || "",





      score:

        Number(item.score) || 0,





      skills:


        Array.isArray(item.skills)

        ? item.skills

        : typeof item.skills === "string"

        ? item.skills.split(",").filter(Boolean)

        : [],





      missingKeywords:


        Array.isArray(item.missingKeywords)

        ? item.missingKeywords


        : Array.isArray(item.missing_keywords)

        ? item.missing_keywords


        : typeof item.missing_keywords === "string"

        ? item.missing_keywords.split(",").filter(Boolean)


        : [],





      strengths:


        Array.isArray(item.strengths)

        ? item.strengths


        : typeof item.strengths === "string"

        ? item.strengths.split(",").filter(Boolean)


        : [],





      suggestions:


        Array.isArray(item.suggestions)

        ? item.suggestions


        : typeof item.suggestions === "string"

        ? item.suggestions.split(",").filter(Boolean)


        : [],





      created_at:

        item.created_at || "",



    })

  );


}